src/table/use-sticky-columns.ts

```diff
@@ -0,0 +1,362 @@
+import { useResizeObserver } from '../internal/hooks/container-queries';
+
+export const selectionColumnId = Symbol('selection-column-id');
+
+// We allow the table to have a minimum of 148px of available space besides the sum of the widths of the sticky columns
+// This value is an UX recommendation and is approximately 1/3 of our smallest breakpoint (465px)
+const MINIMUM_SCROLLABLE_SPACE = 148;
+
+type ColumnId = string | symbol;
+
+interface StickyColumnsProps {
+  visibleColumns: readonly ColumnId[];
+  stickyColumnsFirst: number;
+  stickyColumnsLast: number;
+}
+
+export interface StickyColumnsModel {
+  isEnabled: boolean;
+  store: StickyColumnsStore;
+  style: {
+    wrapper?: React.CSSProperties;
+  };
+  refs: {
+    table: React.RefCallback<HTMLElement>;
+    wrapper: React.RefCallback<HTMLElement>;
+    cell: (columnId: ColumnId, node: null | HTMLElement) => void;
+  };
+}
+
+export interface StickyColumnsState {
+  cellState: Record<ColumnId, null | StickyColumnsCellState>;
+  scrollPadding: ScrollPadding;
+}
+
+// Cell state is used to apply respective styles and offsets to sticky cells.
+export interface StickyColumnsCellState {
+  padLeft: boolean;
+  lastLeft: boolean;
+  lastRight: boolean;
+  offset: { left?: number; right?: number };
+}
+
+// Scroll padding is applied to table's wrapper so that the table scrolls when focus goes behind sticky column.
+export interface ScrollPadding {
+  left: number;
+  right: number;
+}
+
+export function useStickyColumns({
+  visibleColumns,
+  stickyColumnsFirst,
+  stickyColumnsLast,
+}: StickyColumnsProps): StickyColumnsModel {
+  const store = useMemo(() => new StickyColumnsStore(), []);
+  const wrapperRef = useRef<HTMLElement>(null) as React.MutableRefObject<null | HTMLElement>;
+  const tableRef = useRef<HTMLElement>(null) as React.MutableRefObject<null | HTMLElement>;
+  const cellsRef = useRef<Record<ColumnId, HTMLElement>>({});
+
+  const hasStickyColumns = stickyColumnsFirst + stickyColumnsLast > 0;
+
+  const updateStickyStyles = useStableEventHandler(() => {
+    if (wrapperRef.current && tableRef.current) {
+      store.updateCellStyles({
+        wrapper: wrapperRef.current,
+        table: tableRef.current,
+        cells: cellsRef.current,
+        visibleColumns,
+        stickyColumnsFirst,
+        stickyColumnsLast,
+      });
+    }
+  });
+
+  useResizeObserver(wrapperRef, updateStickyStyles);
+
+  useResizeObserver(tableRef, updateStickyStyles);
+
+  useEffect(() => {
+    if (wrapperRef.current && tableRef.current) {
+      store.updateCellStyles({
+        wrapper: wrapperRef.current,
+        table: tableRef.current,
+        cells: cellsRef.current,
+        visibleColumns,
+        stickyColumnsFirst,
+        stickyColumnsLast,
+      });
+    }
+  }, [store, stickyColumnsFirst, stickyColumnsLast, visibleColumns]);
+
+  // Update wrapper styles imperatively to avoid unnecessary re-renders.
+  useEffect(() => {
+    if (!hasStickyColumns) {
+      return;
```

Comment:
1. This condition: https://github.com/cloudscape-design/components/pull/1026/files#diff-9cfc241d7a1172294b6c893446be0a43d07609300fb1db65ad76c0855df90d56R251 ensures one more update is fired after the state changes which should happen before unsubscribe (can check it though);
2. Both hooks also return styles as React props so that the reset happens within the render when the state changes.